src/Disks/IO/WriteBufferFromAzureBlobStorage.cpp

```diff
@@ -77,62 +103,70 @@ void WriteBufferFromAzureBlobStorage::execWithRetry(std::function<void()> func,
+
+
     execWithRetry([this](){ next(); }, max_unexpected_write_error_retries);
 
-    if (tmp_buffer_write_offset > 0)
-        uploadBlock(tmp_buffer->data(), tmp_buffer_write_offset);
+    task_tracker->waitAll();
 
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
     execWithRetry([&](){ block_blob_client.CommitBlockList(block_ids); }, max_unexpected_write_error_retries);
 
     LOG_TRACE(log, "Committed {} blocks for blob `{}`", block_ids.size(), blob_path);
 }
 
-void WriteBufferFromAzureBlobStorage::uploadBlock(const char * data, size_t size)
+void WriteBufferFromAzureBlobStorage::nextImpl()
 {
-    auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
-    const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
-
-    Azure::Core::IO::MemoryBodyStream memory_stream(reinterpret_cast<const uint8_t *>(data), size);
-    execWithRetry([&](){ block_blob_client.StageBlock(block_id, memory_stream); }, max_unexpected_write_error_retries, size);
-    tmp_buffer_write_offset = 0;
-
-    LOG_TRACE(log, "Staged block (id: {}) of size {} (blob path: {}).", block_id, size, blob_path);
+    task_tracker->waitIfAny();
+    writePart();
+    allocateBuffer();
 }
 
-WriteBufferFromAzureBlobStorage::MemoryBufferPtr WriteBufferFromAzureBlobStorage::allocateBuffer() const
+void WriteBufferFromAzureBlobStorage::allocateBuffer()
 {
-    return std::make_unique<Memory<>>(max_single_part_upload_size);
+    buffer_allocation_policy->nextBuffer();
+    auto size = buffer_allocation_policy->getBufferSize();
+    memory = Memory(size);
+    WriteBuffer::set(memory.data(), memory.size());
 }
 
-void WriteBufferFromAzureBlobStorage::nextImpl()
+void WriteBufferFromAzureBlobStorage::writePart()
 {
-    size_t size_to_upload = offset();
+    std::shared_ptr<PartData> part_data;
+    auto data_size = size_t(position() - memory.data());
+    part_data = std::make_shared<PartData>(std::move(memory), data_size);
+    WriteBuffer::set(nullptr, 0);
 
-    if (size_to_upload == 0)
+    if (part_data->data_size == 0)
         return;
 
-    if (!tmp_buffer)
-        tmp_buffer = allocateBuffer();
-
-    size_t uploaded_size = 0;
-    while (uploaded_size != size_to_upload)
+    auto upload_worker = [&, part_data] ()
     {
-        size_t memory_buffer_remaining_size = max_single_part_upload_size - tmp_buffer_write_offset;
-        if (memory_buffer_remaining_size == 0)
-            uploadBlock(tmp_buffer->data(), tmp_buffer->size());
-
-        size_t size = std::min(memory_buffer_remaining_size, size_to_upload - uploaded_size);
-        memcpy(tmp_buffer->data() + tmp_buffer_write_offset, working_buffer.begin() + uploaded_size, size);
-        uploaded_size += size;
-        tmp_buffer_write_offset += size;
-    }
+        auto block_blob_client = blob_container_client->GetBlockBlobClient(blob_path);
+        const std::string & block_id = block_ids.emplace_back(getRandomASCIIString(64));
```

Comment:
This will cause a race. `block_ids` is a shared vector between multiple threads and those threads in your code will try to update `block_ids` without any synchronization. This is bad. Why don't generate `block_id` and put it to `block_ids` BEFORE making `upload_worker`?

Reply:
addressed it